File: server/src/geocode.js
'use strict';

const { db } = require('./db');

/**
 * §4E Cluster Map geocoding. Free OSM/Nominatim by default; provider is
 * swappable (config key `geocode_provider`) in case a paid API key gets
 * added later. Every resolved (or failed) query is cached in
 * `geocode_cache` — a query is never re-sent once attempted.
 *
 * Nominatim's usage policy caps at 1 req/sec and requires an identifying
 * User-Agent, so callers must geocode through `geocodeBatch`, which paces
 * itself, rather than firing requests in parallel.
 *
 * Nominatim frequently can't resolve a full "house number + street" civic
 * address for small/rural Ontario towns (verified while building this —
 * see README) but resolves the street or the town just fine. So each row
 * tries progressively coarser queries and keeps the first hit:
 *   1. full address + municipality
 *   2. street only (house number stripped) + municipality
 *   3. municipality alone (falls back to the town centroid)
 * `precision` on the result records which level actually matched, so the
 * map/heatmap can show that a pin is town-level, not a fabricated exact
 * point.
 */

const NOMINATIM_URL = 'https://nominatim.openstreetmap.org/search';
const USER_AGENT = 'ios-market-tracker/0.1 (local personal-use dashboard)';
const RATE_LIMIT_MS = 1100;

// Strip parenthetical annotations we ourselves added to municipality names
// during import corrections (e.g. "Ancaster (Hamilton)", "New Hamburg
// (Wilmot Twp, Waterloo Region)") — Nominatim chokes on these otherwise.
function cleanMunicipality(municipality) {
  if (!municipality) return municipality;
  return municipality.replace(/\s*\([^)]*\)\s*/g, ' ').replace(/\s+/g, ' ').trim();
}

function stripHouseNumber(address) {
  return address.replace(/^\s*\d+[a-zA-Z]?\s*[-&]?\s*\d*\s*/, '').trim();
}

function buildCandidateQueries(address, municipality) {
  const muni = cleanMunicipality(municipality);
  const queries = [];
  if (address) queries.push({ q: [address, muni, 'Ontario', 'Canada'].filter(Boolean).join(', '), precision: 'address' });
  const street = address ? stripHouseNumber(address) : null;
  if (street && street !== address) queries.push({ q: [street, muni, 'Ontario', 'Canada'].filter(Boolean).join(', '), precision: 'street' });
  if (muni) queries.push({ q: [muni, 'Ontario', 'Canada'].filter(Boolean).join(', '), precision: 'municipality' });
  return queries;
}

function getCached(query) {
  return db.prepare('SELECT * FROM geocode_cache WHERE query = ?').get(query);
}

function setCached(query, lat, lon, precision) {
  db.prepare('INSERT OR REPLACE INTO geocode_cache (query, lat, lon, provider) VALUES (?, ?, ?, ?)').run(
    query, lat, lon, `nominatim:${precision}`
  );
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function nominatimSearch(q) {
  const url = `${NOMINATIM_URL}?format=json&limit=1&countrycodes=ca&q=${encodeURIComponent(q)}`;
  const res = await fetch(url, { headers: { 'User-Agent': USER_AGENT } });
  if (!res.ok) return null;
  const results = await res.json();
  if (results.length === 0) return null;
  return { lat: parseFloat(results[0].lat), lon: parseFloat(results[0].lon) };
}

/** Geocode one address, trying progressively coarser queries; paces itself. */
async function geocodeAddress(address, municipality) {
  const candidates = buildCandidateQueries(address, municipality);
  for (const { q, precision } of candidates) {
    const cached = getCached(q);
    if (cached) {
      if (cached.lat != null) return { lat: cached.lat, lon: cached.lon, precision };
      continue; // this exact query is known to fail — try the next, coarser one
    }
    let result = null;
    try {
      result = await nominatimSearch(q);
    } catch (err) {
      console.error('Geocode request failed for', q, err.message);
    }
    setCached(q, result?.lat ?? null, result?.lon ?? null, precision);
    await sleep(RATE_LIMIT_MS);
    if (result) return { ...result, precision };
  }
  return { lat: null, lon: null, precision: null };
}

/** Geocode up to `limit` listings that don't have lat/lon yet. */
async function geocodeBatch(limit = 20) {
  const rows = db
    .prepare('SELECT source_key, address, municipality FROM listings WHERE lat IS NULL AND address IS NOT NULL LIMIT ?')
    .all(limit);

  let resolved = 0, failed = 0;
  for (const row of rows) {
    const result = await geocodeAddress(row.address, row.municipality);
    if (result.lat != null) {
      db.prepare('UPDATE listings SET lat = ?, lon = ?, geocode_precision = ? WHERE source_key = ?').run(
        result.lat, result.lon, result.precision, row.source_key
      );
      resolved++;
    } else {
      // Permanently give up on this row so it stops being retried forever —
      // it's excluded from "needs geocoding" via geocode_attempted, and the
      // map/heatmap simply won't have a point for it.
      db.prepare('UPDATE listings SET geocode_attempted = 1 WHERE source_key = ?').run(row.source_key);
      failed++;
    }
  }
  const remaining = db.prepare('SELECT COUNT(*) as n FROM listings WHERE lat IS NULL AND geocode_attempted = 0').get().n;
  return { processed: rows.length, resolved, failed, remaining };
}

let backgroundRunning = false;

/**
 * Keep geocoding in the background until every row has a lat/lon (or has
 * permanently failed to resolve). Safe to call as often as you like — a
 * second call while one is already running is a no-op, so "geocode after
 * every import" + "geocode on server startup" never race each other.
 */
async function ensureBackgroundGeocoding() {
  if (backgroundRunning) return;
  backgroundRunning = true;
  try {
    let result;
    do {
      result = await geocodeBatch(20);
    } while (result.processed > 0 && result.remaining > 0);
  } catch (err) {
    console.error('Background geocoding stopped early:', err.message);
  } finally {
    backgroundRunning = false;
  }
}

module.exports = { geocodeBatch, buildCandidateQueries, ensureBackgroundGeocoding };

File: streamlit_app/lib/geocode.py
"""Cluster-map geocoding — ported from geocode.js. Free OSM/Nominatim,
cached forever in geocode_cache. Nominatim frequently can't resolve a full
house-number address for small/rural Ontario towns (verified while
building the Node version) but resolves the street or the town just fine,
so each row tries progressively coarser queries and keeps the first hit.

Streamlit has no real background-worker concept, so geocoding here runs
synchronously in small batches driven by the Cluster Map page's own
"Geocode remaining" button + a progress bar, rather than an always-on
background loop.
"""

import re
import time
import requests

from . import db

NOMINATIM_URL = "https://nominatim.openstreetmap.org/search"
USER_AGENT = "ios-market-tracker-streamlit/0.1 (local personal-use dashboard)"
RATE_LIMIT_SECONDS = 1.1

_PAREN_RE = re.compile(r"\s*\([^)]*\)\s*")
_HOUSE_NUMBER_RE = re.compile(r"^\s*\d+[a-zA-Z]?\s*[-&]?\s*\d*\s*")


def clean_municipality(municipality):
    if not municipality:
        return municipality
    return re.sub(r"\s+", " ", _PAREN_RE.sub(" ", municipality)).strip()


def strip_house_number(address):
    return _HOUSE_NUMBER_RE.sub("", address).strip()


def build_candidate_queries(address, municipality):
    muni = clean_municipality(municipality)
    queries = []
    if address:
        queries.append({"q": ", ".join(filter(None, [address, muni, "Ontario", "Canada"])), "precision": "address"})
    street = strip_house_number(address) if address else None
    if street and street != address:
        queries.append({"q": ", ".join(filter(None, [street, muni, "Ontario", "Canada"])), "precision": "street"})
    if muni:
        queries.append({"q": ", ".join(filter(None, [muni, "Ontario", "Canada"])), "precision": "municipality"})
    return queries


def _get_cached(query):
    return db.query_one("SELECT * FROM geocode_cache WHERE query = ?", (query,))


def _set_cached(query, lat, lon, precision):
    db.execute(
        "INSERT OR REPLACE INTO geocode_cache (query, lat, lon, provider) VALUES (?, ?, ?, ?)",
        (query, lat, lon, f"nominatim:{precision}"),
    )


def _nominatim_search(q):
    url = f"{NOMINATIM_URL}?format=json&limit=1&countrycodes=ca&q={requests.utils.quote(q)}"
    res = requests.get(url, headers={"User-Agent": USER_AGENT}, timeout=10)
    if not res.ok:
        return None
    results = res.json()
    if not results:
        return None
    return {"lat": float(results[0]["lat"]), "lon": float(results[0]["lon"])}


def geocode_address(address, municipality):
    for candidate in build_candidate_queries(address, municipality):
        q, precision = candidate["q"], candidate["precision"]
        cached = _get_cached(q)
        if cached is not None:
            if cached["lat"] is not None:
                return {"lat": cached["lat"], "lon": cached["lon"], "precision": precision}
            continue  # this exact query is known to fail — try the next, coarser one
        result = None
        try:
            result = _nominatim_search(q)
        except requests.RequestException as err:
            print(f"Geocode request failed for {q}: {err}")
        _set_cached(q, result["lat"] if result else None, result["lon"] if result else None, precision)
        time.sleep(RATE_LIMIT_SECONDS)
        if result:
            return {**result, "precision": precision}
    return {"lat": None, "lon": None, "precision": None}


def geocode_batch(limit=20, progress_cb=None):
    """Geocode up to `limit` listings that don't have lat/lon yet.
    progress_cb(done, total) is called after each row, if given.
    """
    rows = db.query(
        "SELECT source_key, address, municipality FROM listings WHERE lat IS NULL AND address IS NOT NULL LIMIT ?",
        (limit,),
    )
    resolved = failed = 0
    for i, row in enumerate(rows):
        result = geocode_address(row["address"], row["municipality"])
        if result["lat"] is not None:
            db.execute(
                "UPDATE listings SET lat = ?, lon = ?, geocode_precision = ? WHERE source_key = ?",
                (result["lat"], result["lon"], result["precision"], row["source_key"]),
            )
            resolved += 1
        else:
            db.execute("UPDATE listings SET geocode_attempted = 1 WHERE source_key = ?", (row["source_key"],))
            failed += 1
        if progress_cb:
            progress_cb(i + 1, len(rows))
    remaining = db.query_one(
        "SELECT COUNT(*) as n FROM listings WHERE lat IS NULL AND geocode_attempted = 0"
    )["n"]
    return {"processed": len(rows), "resolved": resolved, "failed": failed, "remaining": remaining}
